Add option for a backend check to approve use of on-demand TLS

tantalic opened the pull request:
### 1. What does this change do, exactly?
This adds the ask sub-directive to tls that defines the URL of a backend HTTP service to be queried during the TLS handshake to determine if an on-demand TLS certificate should be acquired for incoming hostnames. When the ask sub-directive is defined, Caddy will query the URL for permission to acquire a cert by making a HTTP GET request to the URL including the requested domain in the query string. If the backend service returns a 2xx response Caddy will acquire a cert. Any other response code (including 3xx redirects) are be considered a rejection and the certificate will not be acquired.

### 2. Please link to the relevant issues.
#1847

### 3. Which documentation changes (if any) need to be made because of this PR?
The TLS docs will need a reference to the `ask` directive and some updated language around on-demand TLS in general. 

### 4. Checklist

- [ ] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I am willing to help maintain this change if there are issues with it later

tantalic commented on caddytls/handshake.go:
```diff
@@ -159,10 +161,49 @@ func (cfg *Config) getCertDuringHandshake(name string, loadIfNecessary, obtainIf
+
+func (cfg *Config) checkURLForObtainingNewCerts(name string) error {
+	client := http.Client{
+		Timeout: 10 * time.Second,
```
I set a timeout of 10s here. Since this should be a backend service that should be able to respond quickly and it will be holding open the TLS negotiation I think we could drop this quite a bit lower as well. It could be configurable as well, but that seems unnecessary to me at this time.

mholt commented on caddytls/setup.go:
```diff
@@ -164,6 +164,9 @@ func setupTLS(c *caddy.Controller) error {
 				c.Args(&maxCerts)
 				config.OnDemand = true
+			case "ask":
+				c.Args(&askURL)
```
See if you can parse the URL here, and if it fails, return an error. This will help prevent configuration errors after the server has already started...

tantalic replied:
See line 220 in the latest revision which validates the URL can be parsed and the scheme is either HTTP or HTTPS.

mholt commented on caddytls/handshake.go:
```diff
@@ -159,10 +161,49 @@ func (cfg *Config) getCertDuringHandshake(name string, loadIfNecessary, obtainIf
+	}
+
+	if resp.StatusCode < 200 || resp.StatusCode > 299 {
+		return fmt.Errorf("Non-200 status code returned (%d)", resp.StatusCode)
```
Start the error string with a lower-case letter, since it is usually used in the middle of a sentence when printing to the logs. Also, be a little more descriptive about the implications of this. For example: "certificate for hostname '%s' not allowed, non-2xx status code %d returned"

mholt commented on caddytls/config.go:
```diff
@@ -148,6 +148,11 @@ type OnDemandState struct {
+	// The url to call to check if an on-demand tls certificate should
+	// be issued. If a request to the URL fails or returns a non 2xx
+	// status on-demand issuances must fail.
+	CheckUrl string
```
"CheckURL" or, for consistency with the Caddyfile, "AskURL" -- or for more descriptive: "OnDemandAskURL" or something.

tantalic commented:
@mholt thanks for the review, this now includes the improvements requested.

mholt commented on caddytls/handshake.go:
```diff
@@ -159,10 +161,48 @@ func (cfg *Config) getCertDuringHandshake(name string, loadIfNecessary, obtainIf
+	if err != nil {
+		return err
+	}
+
```
The response body must be closed:

```go
defer resp.Body.Close()
```

tobya commented:
Is there any benefit of thinking about this as an event for on? Eg

    On tls_anyhost https://example.com

Just an idea

tantalic commented:
@mholt this has been updated as follows:

- Added the missing `resp.Body.Close()` as requested
- Updated the backend HTTP request so that it doesn't follow 3xx redirects (`handshake.go` lines 181-183), this matches the discussion in  #1847

mholt commented:
LGTM. Thank you!!
